src/Functions/LowerUpperImpl.h

```diff
@@ -1,99 +1,195 @@
 #pragma once
 #include <Columns/ColumnString.h>
+#include <Common/TargetSpecific.h>
+
+#if USE_MULTITARGET_CODE
+#include <immintrin.h>
+#endif
 
 
 namespace DB
 {
 
+DECLARE_DEFAULT_CODE(
 template <char not_case_lower_bound, char not_case_upper_bound>
-struct LowerUpperImpl
+static void arrayImpl(const UInt8 * src, const UInt8 * src_end, UInt8 * dst)
 {
-    static void vector(const ColumnString::Chars & data,
-        const ColumnString::Offsets & offsets,
-        ColumnString::Chars & res_data,
-        ColumnString::Offsets & res_offsets)
-    {
-        res_data.resize(data.size());
-        res_offsets.assign(offsets);
-        array(data.data(), data.data() + data.size(), res_data.data());
-    }
+    static constexpr auto flip_case_mask = 'A' ^ 'a';
+     for (; src < src_end; ++src, ++dst)
+        if (*src >= not_case_lower_bound && *src <= not_case_upper_bound)
+            *dst = *src ^ flip_case_mask;
+        else
+            *dst = *src;
+})
+
+DECLARE_AVX512BW_SPECIFIC_CODE(
+template <char not_case_lower_bound, char not_case_upper_bound>
+static void arrayImpl(const UInt8 * src, const UInt8 * src_end, UInt8 * dst)
+{
+    static constexpr auto flip_case_mask = 'A' ^ 'a';
 
-    static void vectorFixed(const ColumnString::Chars & data, size_t /*n*/, ColumnString::Chars & res_data)
+    const auto byte_avx512 = sizeof(__m512i);
+    const auto src_end_avx = src_end - (src_end - src) % byte_avx512;
+    if (src < src_end_avx)
     {
-        res_data.resize(data.size());
-        array(data.data(), data.data() + data.size(), res_data.data());
+        const auto v_not_case_lower_bound = _mm512_set1_epi8(not_case_lower_bound - 1);
+        const auto v_not_case_upper_bound = _mm512_set1_epi8(not_case_upper_bound + 1);
+
+        for (; src < src_end_avx; src += byte_avx512, dst += byte_avx512)
+        {
+            const auto chars = _mm512_loadu_si512(reinterpret_cast<const __m512i *>(src));
```

Comment:
I know, it has been like that before but adding comments similar to l. 77, 80, etc. would be nice for consistency.